OSDOCS-7772: Migrated the GitOps with ArgoCD topic (#67145)
openshift/openshift-docs: cloud_experts_tutorials/cloud-experts-gitops-with-argocd.adoc

:_mod-docs-content-type: ASSEMBLY
[id=“cloud-experts-gitops-with-argocd]

> = Tutorial: Demonstrate GitOps on Managed OpenShift with ArgoCD

Reviewer: ```suggestion
= Tutorial: Demonstrate GitOps on Managed OpenShift with Argo CD
```

include::_attributes/attributes-openshift-dedicated.adoc[]
:context: cloud-experts-gitops-with-argocd

toc::[]

//Mobb content metadata
//Brought into ROSA product docs 2023-10-31
// ---
// date: '2021-06-17'
// title: Demonstrate GitOps on Managed OpenShift with ArgoCD
// aliases: ['/experts/demos/gitops/']
// tags: ["GitOps", "ROSA", "ARO"]
// authors:
//   - Steve Mirman
//   - Connor Wooley
// ---

> The purpose of this document is to help you get  OpenShift GitOps running in your cluster, including deploying a sample application and demonstrating how ArgoCD ensures environment consistency.

Reviewer: ```suggestion
This tutorial helps you run OpenShift GitOps in your cluster, deploy a sample application, and demonstrates how Argo CD ensures environment consistency.
```

[NOTE]
====

> If you prefer visual guide, you can watch a quick start video on link:https://www.youtube.com/watch?v=Gi18iemF1yI[YouTube].

Reviewer: ```suggestion
If you prefer a visual guide, you can watch a quick start video on link:https://www.youtube.com/watch?v=Gi18iemF1yI[YouTube].
```

====

== Prerequisites

* You have a Managed OpenShift Cluster with cluster-admin rights.
* You have the following tools installed:

>     ** link:https://git-scm.com/download/[git]

Reviewer: ```suggestion
** link:https://git-scm.com/download/[Git]
```
I think you can delete the spaces before the bullets (https://docs.asciidoctor.org/asciidoc/latest/lists/unordered/).

>     ** link:https://docs.openshift.com/container-platform/4.13/cli_reference/openshift_cli/getting-started-cli.html#cli-installing-cli_cli-developer-commands[oc]

Reviewer: ```suggestion
** link:https://docs.openshift.com/container-platform/4.13/cli_reference/openshift_cli/getting-started-cli.html#cli-installing-cli_cli-developer-commands[OpenShift command-line interface (CLI) `oc`]
```

2. Clone the `gitops-demo` GitHub repository to your local machine by running the following command:
+
[source,terminal]
----
$ git clone https://github.com/rh-mobb/gitops-demo gitops
----

3. Export your local path to the GitHub files:
+
[source,terminal]
----
$ export GITOPS_HOME="$(pwd)/gitops"
$ cd $GITOPS_HOME
----

4. Log in to OpenShift by copying the login command from the OpenShift console:
+
image::oc_login.png[screenshot of login]

5. Enter the command in your terminal to authenticate with the OpenShift CLI (oc).
+
.Sample output
+
[source,terminal]
----
Logged into "https://<YOUR-INSTANCE>.openshiftapps.com:6443" as "<YOUR-ID>" using the token provided.
----

== Deploy the ArgoCD Project

1. Create a new OpenShift project called *gitops* by running the following command:
+
[source,terminal]
----
$ oc new-project gitops
----

2. Add **cluster-admin** rights to the `openshift-gitops-argocd-application-controller` service account in the **openshift-gitops** namespace by using the following command:
+
[source,terminal]
----
$ oc adm policy add-cluster-role-to-user cluster-admin -z openshift-gitops-argocd-application-controller -n openshift-gitops
----

> 3. Retrieve the ArgoCD URL to log into your ArgoCD:

Reviewer: ```suggestion
. Retrieve the Argo CD URL to log in to your Argo CD:
```

+
[source,terminal]
----
$ argoURL=$(oc get route openshift-gitops-server -n openshift-gitops -o jsonpath='{.spec.host}{"\n"}')
$ echo $argoURL
----

> 4. Retrieve ArgoCD Password:

Reviewer: ```suggestion
. Retrieve the Argo CD password:
```

+
[source,terminal]
----
$ argoPass=$(oc get secret/openshift-gitops-cluster -n openshift-gitops -o jsonpath='{.data.admin\.password}' | base64 -d)
$ echo $argoPass
----

> 5. In a web browser, navigate to the ArgoCD console using the `$argoURL` value:

Reviewer: ```suggestion
. In a web browser, navigate to the Argo CD console using the `$argoURL` value:
```

+
image::argo1.png[screenshot of argocd1]

> 6. Log in with the user name **admin** and the password returned as `$argoPass`:

Reviewer: ```suggestion
. Log in with the username *admin* and the password returned as `$argoPass`:
```
I'm not sure why the username is bold and the password is in code block here. I would think it would be one or the other?

+
image::argo2.png[screenshot of argocd2]

> 7. Optional: if you want to use the command-line interface, log in to the CLI by running the following command:

Reviewer: ```suggestion
. Optional: if you want to use the Argo CD CLI, run the following command:
```

+
[source,terminal]
----
$ argocd login --insecure --grpc-web $argoURL  --username admin --password $argoPass
----

> 8. Use the `kubectl` CLI tool to apply the `bgd-app.yaml` file

Reviewer: ```suggestion
. Use the `kubectl` CLI tool to apply the `bgd-app.yaml` file
```
Is this the same as the `kam CLI` referenced in line 35? Should we use the same name throughout?

Author: Not sure - the link goes to `Kustomize` which is a separate level from `kubectl` - https://kubectl.docs.kubernetes.io/installation/kustomize/

+
[source,terminal]
----
$ kubectl apply -f documentation/modules/ROOT/examples/bgd-app/bgd-app.yaml
----
+

> The `bgd-app.yaml` file defines several things, including the repository's location for the `gitops-bgd-app` application:

Reviewer: I really don't like the use of "things" here, but it might be better to fix on the QE round.

+
image::bgd-app-yaml.png[screenshot of bgd-app-yaml]

> 9. Check the rollout running the following command:

Reviewer: ```suggestion
. Check the rollout by running the following command:
```

+
[source,terminal]
----
$ kubectl rollout status deploy/bgd -n bgd
----

> 10. Once the rollout is completed, get the route to the application by running the following command:

Reviewer: ```suggestion
. Once the rollout is complete, obtain the route to the application by running the following command:
```

+
[source,terminal]
----
$ oc get route bgd -n bgd -o jsonpath='{.spec.host}{"\n"}'
----

> 11. In your browser, paste the route to open the application:

Reviewer: ```suggestion
. In your browser, paste the route to open the application:
```

+
image::app_blue.png[screenshot of app_blue]

> 12. Go back to your ArgoCD window and verify the configuration shows:

Reviewer: ```suggestion
. Go back to your Argo CD window and verify that the configuration shows:
```
I really don't like ending with "shows" here. Shows what? But it might be something to fix on the second round.

+
image::argo_app1.png[screenshot of argo_app1]

> 13. Exploring the application in ArgoCD, you can see all the components are green, which denotes that these components are synchronized:

Reviewer: ```suggestion
. Explore the application in Argo CD. All the components are green, which denotes that these components are synchronized:
```

+
image::argo_sync.png[screenshot of argo_sync]

> == Deploy a change to the application

Reviewer: ```suggestion
== Deploying a change to the application
```

> 1. In the terminal, enter the following command, which introduces a change into the bgd application:

Reviewer: ```suggestion
. In the terminal, enter the following command, which introduces a change into the bgd application:
```

+
[source,terminal]
----
$ kubectl -n bgd patch deploy/bgd --type='json' -p='[{"op": "replace", "path": "/spec/template/spec/containers/0/env/0/value", "value":"green"}]'
----

> 2. In your ArgoCD window, the application should no longer be synchronized:

Reviewer: ```suggestion
. In your Argo CD window, the application should no longer be synchronized:
```

+
image::argo_out_of_sync.png[screenshot of argo_sync]

> 3. Refresh the bgd application window and notice the change in box color:

Reviewer: ```suggestion
. Refresh the bgd application window and note the change in box color:
```

+
image::bgd_green.png[screenshot of bgd_green]
+

> The new deployment changed the box from blue to green, but only within OpenShift, not in the source code repository

Reviewer: ```suggestion
The new deployment changed the box from blue to green. However, the change is only in the OpenShift repository and not in the source code repository.
```

> == Synchronize the application

Reviewer: ```suggestion
== Synchronizing an application
```

> 1. In the ArgoCD console, click the `SYNC` button to re-synchronize the bgd application with the approved configuration in the source code repository:

Reviewer: ```suggestion
. In the Argo CD console, click the `SYNC` button to re-synchronize the bgd application with the approved configuration in the source code repository:
```

+
image::sync_bgd.png[screenshot of sync_bgd]

> 2. Refresh the bgd application window and notice the box color change:

Reviewer: ```suggestion
. Refresh the bgd application window and note the box color change:
```
